File: pyvalerie/src/GoogleProvider.py
# -*- coding: utf-8 -*-

import re

import WebGrabber

from Plugins.Extensions.ProjectValerie.__common__ import printl2 as printl

#------------------------------------------------------------------------------------------

class GoogleProvider():

	URL = u"http://www.google.com/"
	apiSearch = URL + u"search?q=<search>"
	
	# Issue #205, efo => don't use start tag
	# DIV_RESULT_START = u"<div id=\"search\">"
	DIV_RESULT_FLAG = u"<h3 class=\"r\">"

	def searchForSeasonAndEpisode(self, info, result):
		m = re.search(r's(?P<season>\d+)e(?P<episode>\d+)', result)
		if m and m.group("season") and m.group("episode"):
			printl("m.group()=" + str(m.group()), self)
			info.Season = int(m.group("season"))
			info.Episode = int(m.group("episode"))
			return info
		
		m = re.search(r'(?P<season>\d+)x(?P<episode>\d+)', result)
		if m and m.group("season") and m.group("episode"):
			printl("m.group()=" + str(m.group()), self)
			info.Season = int(m.group("season"))
			info.Episode = int(m.group("episode"))
			return info
		
		return None

	def getSeasonAndEpisodeFromEpisodeName(self, info):
		if info.SearchString is None or len(info.SearchString) == 0:
			printl(" <- None (info.SearchString is None or len(info.SearchString) == 0)", self) 
			return (False, None)
		
		url = self.apiSearch
		url = re.sub("<search>", info.SearchString, url)
		html = WebGrabber.getHtml(url)
		
		if html is None:
			printl(" <- None (html is None)", self) 
			return (False, None)
		
		# well there seems to be a problem with detecting tvshows,
		#so lets build in a workaround, you will need at least 2 time the same
		#season and episode before acepting it
		
		count = 0
		s = 0
		e = 0
		
		# Issue #205, efo => use only result tag
		# htmlSplitted = html.split(self.DIV_RESULT_START)
		htmlSplitted = html.split(self.DIV_RESULT_FLAG)
		lobSkipEntry = True
		for htmlSplitter in htmlSplitted:
			# Skip first entry, since this mostly contain script stuff...
			if lobSkipEntry == True:
				lobSkipEntry = False
				continue
			
			htmlSplitter = htmlSplitter.strip()
			
			# Issue #205, efo => don't split again, since we now already have a single line which can be checked...
			#if htmlSplitter.startswith(self.DIV_RESULT_FLAG) is False:
			#	continue
			
			#pos = htmlSplitter.find(self.DIV_RESULT_END)
			#if pos < 0:
			#	continue
			
			tmp = self.searchForSeasonAndEpisode(info, htmlSplitter.lower())
			if tmp is not None:
				info = tmp
				
				if s == 0 or e == 0:
					s = info.Season
					e = info.Episode

				if s == info.Season and e == info.Episode:
					count = count + 1
					if count == 2:
						return (True, info)
			else:
				continue
		
		printl(" <- None (eof)", self) 
		return (True, None)

File: pc/src/valerie/provider/GoogleProvider.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */

package valerie.provider;

import java.util.regex.Matcher;
import java.util.regex.Pattern;
import valerie.MediaInfo;
import valerie.tools.DebugOutput;

/**
 *
 * @author i7
 */
public class GoogleProvider {

    private static final String URL = "http://www.google.com/";
    private static final String apiSearch = URL + "search?q=<search>";


    private static final String DIV_RESULT_START = "<li class=g>";
    private static final String DIV_RESULT_FLAG = "<h3 class=\"r\">";


    private boolean searchForSeasonAndEpisode(MediaInfo info, String result) {
        Matcher m = Pattern.compile("s\\d+e\\d+").matcher(result);
        if (m.find()) {
            String group = m.group();
            m = Pattern.compile("s\\d+").matcher(group.trim());
            if (m.find()) {
                info.Season = Integer.valueOf(m.group().substring(1).trim());
            }
            m = Pattern.compile("e\\d+").matcher(group.trim());
            if (m.find()) {
                info.Episode = Integer.valueOf(m.group().substring(1).trim());
            }
            return true;
        }
        return false;
    }

    public boolean getSeasonAndEpisodeFromEpisodeName(MediaInfo info) {
        DebugOutput.printl("");

        if(info.SearchString.length() <= 0)
            return false;

        String html = null;

        String url = apiSearch;
        url = url.replaceAll("<search>", info.SearchString);
        html = valerie.tools.WebGrabber.getHtml(url);

        if (html == null)
            return false;

        String[] htmlSplitted = html.split(DIV_RESULT_START);
        for(String htmlSplitter : htmlSplitted) {
            htmlSplitter = htmlSplitter.trim();
            if(!htmlSplitter.startsWith(DIV_RESULT_FLAG))
                continue;

            //int pos = htmlSplitter.indexOf(DIV_RESULT_END);
            //if(pos < 0)
            //    continue;

            if (searchForSeasonAndEpisode(info, htmlSplitter.toLowerCase()))
                return true;
            else
                continue;
        }
        return false;
    }
}
